### PR TITLE
Trigger for time of day

### PR DESCRIPTION
**Pull request by PoweRGbg**
Now creatinng time ranges is easy

### CODE REVIEW

File: app/src/main/java/info/nightscout/androidaps/plugins/general/automation/triggers/TriggerTimeRange.java [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+
+// Trigger for time range ( from 10:00AM till 13:00PM )
+
+public class TriggerTimeRange extends Trigger {
+    private static Logger log = LoggerFactory.getLogger(L.AUTOMATION);
+    
+    // in minutes since midnight 60 means 1AM
+    private int start;
+    private int end;
+    private Comparator comparator = new Comparator();
```
Comment (MilosKozak) [inline app/src/main/java/info/nightscout/androidaps/plugins/general/automation/triggers/TriggerTimeRange.java]: not used

File: app/src/main/java/info/nightscout/androidaps/plugins/general/automation/triggers/TriggerTimeRange.java [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+        return this;
+    }
+
+    @Override
+    public Trigger duplicate() {
+        return new TriggerTimeRange(this);
+    }
+
+    long toMilis(long minutesSinceMidnight) {
+        long hours =  minutesSinceMidnight / 60;//hours
```
Comment (MilosKozak) [inline app/src/main/java/info/nightscout/androidaps/plugins/general/automation/triggers/TriggerTimeRange.java]: isn't this just `* 60 * 1000`?

File: app/src/main/java/info/nightscout/androidaps/plugins/general/automation/triggers/TriggerTimeRange.java (hunk 1 above)
Comment (PoweRGbg) [reply app/src/main/java/info/nightscout/androidaps/plugins/general/automation/triggers/TriggerTimeRange.java]: removed

File: app/src/main/java/info/nightscout/androidaps/plugins/general/automation/triggers/TriggerTimeRange.java (hunk 2 above)
Comment (PoweRGbg) [reply app/src/main/java/info/nightscout/androidaps/plugins/general/automation/triggers/TriggerTimeRange.java]: You are right. Fixed

**Comment by MilosKozak** [comment]
setting time is not working properly

File: app/src/test/java/info/nightscout/androidaps/plugins/general/automation/triggers/TriggerTimeRangeTest.java [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+        TriggerTimeRange t;
+//        when(System.currentTimeMillis()).thenReturn((long) now*60000);
+        when(DateUtil.now()).thenReturn((long) now*60000);
+
+        // range starts 1 min in the future
+        t = new TriggerTimeRange().period(now + 1, now + 30);
+        Assert.assertEquals(false, t.shouldRun());
+
+        // range starts 30 min back
+        t = new TriggerTimeRange().period((int) (System.currentTimeMillis() - 30*60000), now + 30);
```
Comment (MilosKozak) [inline app/src/test/java/info/nightscout/androidaps/plugins/general/automation/triggers/TriggerTimeRangeTest.java]: System.currentTimeMillis()  is not mocked

File: app/src/main/java/info/nightscout/androidaps/plugins/general/automation/triggers/TriggerTimeRange.java [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+
+        startButton.setOnClickListener(view -> {
+            GregorianCalendar calendar = new GregorianCalendar();
+            calendar.setTimeInMillis(toMilis(start));
+            TimePickerDialog tpd = TimePickerDialog.newInstance(
+                    (view12, hourOfDay, minute, second) -> {
+                        calendar.set(Calendar.HOUR_OF_DAY, hourOfDay);
+                        calendar.set(Calendar.MINUTE, minute);
+                        start = getMinSinceMidnight(calendar.getTimeInMillis());
+                        startButton.setText(DateUtil.timeString(start));
```
Comment (MilosKozak) [inline app/src/main/java/info/nightscout/androidaps/plugins/general/automation/triggers/TriggerTimeRange.java]: `startButton.setText(DateUtil.timeString(toMilis(start)));` ?

File: app/src/main/java/info/nightscout/androidaps/plugins/general/automation/triggers/TriggerTimeRange.java [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+        });
+        endButton.setOnClickListener(view -> {
+            GregorianCalendar calendar = new GregorianCalendar();
+            calendar.setTimeInMillis(toMilis(end));
+            TimePickerDialog tpd = TimePickerDialog.newInstance(
+                    (view12, hourOfDay, minute, second) -> {
+                        calendar.set(Calendar.HOUR_OF_DAY, hourOfDay);
+                        calendar.set(Calendar.MINUTE, minute);
+                        end = getMinSinceMidnight(calendar.getTimeInMillis());
+                        endButton.setText(MainApp.gs(R.string.and) + " " + DateUtil.timeString(end));
```
Comment (MilosKozak) [inline app/src/main/java/info/nightscout/androidaps/plugins/general/automation/triggers/TriggerTimeRange.java]: the same here

File: app/src/main/java/info/nightscout/androidaps/plugins/general/automation/triggers/TriggerTimeRange.java (hunk 4 above)
Comment (PoweRGbg) [reply app/src/main/java/info/nightscout/androidaps/plugins/general/automation/triggers/TriggerTimeRange.java]: corrected. Thanks!

**Comment by MilosKozak** [comment]
editing time is still not working properly

File: app/src/main/java/info/nightscout/androidaps/plugins/general/automation/triggers/TriggerTimeRange.java [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+            Activity a = scanForActivity(root.getContext());
+            if (a != null)
+                tpd.show(a.getFragmentManager(), "TimePickerDialog");
+        });
+        endButton.setOnClickListener(view -> {
+            GregorianCalendar calendar = new GregorianCalendar();
+            calendar.setTimeInMillis(toMilis(end));
+            TimePickerDialog tpd = TimePickerDialog.newInstance(
+                    (view12, hourOfDay, minute, second) -> {
+                        calendar.set(Calendar.HOUR_OF_DAY, hourOfDay);
```
Comment (MilosKozak) [inline app/src/main/java/info/nightscout/androidaps/plugins/general/automation/triggers/TriggerTimeRange.java]: I'd start with creating new calendar instance on these places

**Comment by MilosKozak** [comment]
where is this commit from ? 
@PoweRGbg improved statuslights layout (#78)

**Comment by PoweRGbg** [comment]
@MilosKozak No idea where #78 came from, but I reverted it

**Pull request merged**